tests/matrix_free/matrix_vector_faces_24.cc

```diff
@@ -0,0 +1,109 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+
+// check the setup of ghost faces and elements in case the flag
+// MatrixFree::AdditionalData::hold_all_faces_to_owned_cells is set to true
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/fe/fe_dgq.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_tools.h>
+
+#include <deal.II/lac/la_parallel_vector.h>
+
+#include "../tests.h"
+#include "create_mesh.h"
+
+std::ofstream logfile("output");
+
+#include "matrix_vector_faces_common.h"
+
+
+
+template <int dim, int fe_degree>
+void
+test()
+{
+  if (fe_degree > 1)
+    return;
+
+  // raise element degree by two to test cubic and quartic shape functions
+  // rather than linears and quadratics according to
+  // matrix_vector_faces_common.h
+
+  parallel::distributed::Triangulation<dim> tria(MPI_COMM_WORLD);
+  GridGenerator::hyper_cube(tria);
+  tria.refine_global(1);
+  tria.last()->set_refine_flag();
```

Comment:
So with this mesh you could trigger the bug in both 2D and 3D?

Reply:
Yes, the important part is that all 2^dim cells of level 2 sit on one processor, whereas the other 2^dim-1 cells on level 1 sit on the other processor. (The problem did not happen without the `MatrixFree::AdditionalData::hold_all_faces_to_owned_cells`, which was not tested well enough before.)